### PR TITLE
TC: Polish trait validation checking and error reporting

### LINKED ISSUE #389: TC: Properly ensure impls implement their traits, and infer original types while at it
1. Ensure all members in the impl appear in the trait, and have correct types (extension: if types are incomplete, infer from trait. Maybe this should be done in the future). (Unification needs to happen with src=impl and target=trait)
3. Ensure all members in the trait appear in the impl.

### PR DESCRIPTION
**Pull request by feds01**
closes #389

### CODE REVIEW

File: compiler/hash-typecheck/src/diagnostics/error.rs [hunk 1]
```diff
@@ -1189,46 +1189,50 @@ impl<'tc> From<TcErrorWithStorage<'tc>> for Report {
...
+                // Create a sequence display for displaying member names
+                let missing_members = SequenceDisplay::new(
+                    &missing,
+                    SequenceDisplayOptions::with_limit(SequenceJoinMode::All, 6),
+                );
+
                 builder.with_error_code(HashErrorCode::TraitImplMissingMember).with_message(
                     format!(
-                        "trait `{}` is missing the member `{}`",
+                        "trait `{}` is missing the trait item{} {missing_members}",
```
Comment (kontheocharis) [inline compiler/hash-typecheck/src/diagnostics/error.rs:1211]: missing space? Also maybe "member" instead of "trait item"?

**Review by kontheocharis — APPROVED** [review]

File: compiler/hash-typecheck/src/diagnostics/error.rs (hunk 1 above)
Comment (feds01) [reply compiler/hash-typecheck/src/diagnostics/error.rs:1211]: No missing space, `pluralise!` basically converts the number into a suffix for a given word, which means that `1` yields nothing and anything greater yields `s`.

Comment (feds01) [reply compiler/hash-typecheck/src/diagnostics/error.rs:1211]: the two phrases are synonymous so I don't mind which one we use

Comment (kontheocharis) [reply compiler/hash-typecheck/src/diagnostics/error.rs:1211]: Probably member is best for the sake of consistency

**Review by kontheocharis — APPROVED** [review]

**Pull request merged**